File: web-app/src/app/shared/form-utils/leave-confirmation/guard/leave-confirmation.guard.ts
import {Injectable} from '@angular/core';
import {CanDeactivate} from '@angular/router';
import {ComponentLeaveConfirmation} from './component-leave-confirmation';

@Injectable({
    providedIn: 'root'
})
export class LeaveConfirmationGuard implements CanDeactivate<ComponentLeaveConfirmation> {

    canDeactivate(component: ComponentLeaveConfirmation): boolean {
        if(!component.canDeactivate())
            return confirm('If you leave before saving, your changes will be lost.');

        return true;
    }

}

File: application/src/test/java/objective/taskboard/it/components/guards/LeaveConfirmationGuard.java
package objective.taskboard.it.components.guards;

import static objective.taskboard.it.AbstractUiFragment.waitUntil;

import org.openqa.selenium.NoAlertPresentException;
import org.openqa.selenium.WebDriver;

public class LeaveConfirmationGuard {

    private static final String DEFAULT_MESSAGE = "If you leave before saving, your changes will be lost.";

    public static void waitIsOpened(WebDriver webDriver) {
        waitUntil(webDriver, (w) -> {
            return isAlertOpened(webDriver) && DEFAULT_MESSAGE.equals(webDriver.switchTo().alert().getText());
        });
    }

    public static void waitIsClosed(WebDriver webDriver) {
        waitUntil(webDriver, (w) -> !isAlertOpened(webDriver) || !DEFAULT_MESSAGE.equals(webDriver.switchTo().alert().getText()));
    }

    private static boolean isAlertOpened(WebDriver webDriver) {
        try {
            webDriver.switchTo().alert();
            return true;
        } catch (NoAlertPresentException e) {
            return false;
        }
    }

    public static void stay(WebDriver webDriver) {
        waitIsOpened(webDriver);
        webDriver.switchTo().alert().dismiss();
        waitIsClosed(webDriver);
    }

    public static void leave(WebDriver webDriver) {
        waitIsOpened(webDriver);
        webDriver.switchTo().alert().accept();
        waitIsClosed(webDriver);
    }

}
